File: python/person.py
from typing import Tuple, List


class Person:

    def __init__(self, _id: int, state: int) -> None:
        self._id = _id
        self.state = state

        # Nodo padre correspondiente al árbol
        self.parent = None

        # Nodo cabeza y cola correspondiente a la lista ligada-nodos hijos
        self.head = None
        self.tail = None

        # Nodo anterior y siguiente correspondiente a los nodos hermanos
        self.prev = None
        self._next = None 

    def append_contact(self, other_id: int, state: int):
        """ 
        Inserta una nueva persona como nodo hijo al final de la lista ligada
        """
        
        # Nuevo nodo con id dado y padre self
        new = Person(other_id, state)
        new.parent = self
        
        # insertar nodo al final de la lista
        if not self.head:
            self.head = new
        else:
            new.prev = self.tail
            self.tail._next = new
        self.tail = new

        return new

    def search_contact(self, _id: int):
        """ 
        Busca el nodo hijo con id dado 
        """
        current = self.head
        while current:
            if current._id == _id:
                return current
            current = current._next
        return None
    
    def recursive_inform(self, depth: int, output_file):
        output_file.write("    "*depth + f"{self._id}:{self.state}\n")
        current = self.head
        while current:
             current.recursive_inform(depth + 1, output_file)
             current = current._next
        
    def recursive_delete(self):
        current = self.head
        while current:
             current.recursive_delete()
             current = current._next
             if current:
                current._next = None

        
    def discard_person(self):
        self.recursive_delete()
        if self.prev and self._next:
            self.prev._next = self._next
            self._next.prev = self.prev
            
        elif self.prev and not self._next:
            self.prev._next = None
            self.parent.tail = self.prev
            
        elif not self.prev and self._next:
            self._next.prev = None
            self.parent.head = self._next
            
        elif not self.prev and not self._next:
            self.parent.head = None
            self.parent.tail = None
            
    def recursive_statistics(self, stats: List, person=None):
        
        if not person:
             current = self.head            
             stats[self.state] += 1
        else:
            current = person.head
            stats[person.state] += 1
        while current:
             current.recursive_statistics(stats, current)
             current = current._next

File: src/structs/person.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "person.h"

Person* person_init(int id, int state){
    Person* p = (Person*) malloc(sizeof(Person));
    p->id = id;
    p->state = state;
    p->head = NULL;
    p->next = NULL;
    p->parent = NULL;
    p->prev = NULL;
    p->tail = NULL;
    return p;
    
}

Person* person_append_contact(Person* person, int other_id, int state){
    Person* new;
    new = person_init(other_id, state);
    new->parent = person;
    
    if (person->head == NULL){
        person->head = new;
    }
    else{
        new->prev = person->tail;
        person->tail->next = new;
    }
    person->tail = new;
    return new;
};

Person* person_search_contact(Person* person, int id){
    Person* current = person->head;
    while (current != NULL){
        if (current->id == id) return current;
        current = current->next;
    }
    return NULL;
}

void person_recursive_inform(Person* person, int depth, FILE* file){
    char spc_str[4*depth + 1];
    for (int i=0; i<4*depth; i++){
        spc_str[i] = ' ';
    }
    spc_str[4*depth] = '\0';
    fprintf(file, "%s%d:%d\n",spc_str, person->id, person->state);
    Person* current = person->head;
    while (current != NULL){
        person_recursive_inform(current, depth+1, file);
        current = current->next;
    } 
}
void person_recursive_destroy(Person* person){
    Person* current = person->head;
    Person* last = NULL;
    while (current != NULL){
        free(last);
        person_recursive_destroy(current);
        last = current;  
        current = current->next;  
    }
    free(last);
}

void person_discard(Person* person){
    person_recursive_destroy(person);
    if (person->prev && person->next){
        person->prev->next = person->next;
        person->next->prev = person->prev;
    }
            
    else if (person->prev && !person->next){
        person->prev->next = NULL;
        person->parent->tail = person->prev;
    }
            
    else if ( !person->prev && person->next){
        person->next->prev = NULL;
        person->parent->head = person->next;
    }            
    else if (!person->prev && !person->next){
        person->parent->head = NULL;
        person->parent->tail = NULL;
    }
    free(person);
}